lib/real-device-management.js

```diff
@@ -67,20 +68,45 @@ async function runRealDeviceReset(device, opts) {
  * @param {InstallOptions} [opts]
  */
 async function installToRealDevice(device, app, bundleId, opts) {
-  if (!device.udid || !app) {
-    log.debug('No device id or app, not installing to real device.');
+  if (!device.udid || !app || !bundleId) {
+    log.debug('No device id, app or bundle id, not installing to real device.');
     return;
   }
 
   const {skipUninstall, strategy, timeout} = opts ?? {};
 
-  if (!skipUninstall && bundleId && (await device.isAppInstalled(bundleId))) {
-    log.debug(`Reset requested. Removing app with id '${bundleId}' from the device`);
+  if (!skipUninstall) {
+    log.info(`Reset requested. Removing app with id '${bundleId}' from the device`);
     await device.remove(bundleId);
   }
   log.debug(`Installing '${app}' on device with UUID '${device.udid}'...`);
-  await device.install(app, timeout, strategy);
-  log.debug('The app has been installed successfully.');
+
+  try {
+    await device.install(app, timeout, strategy);
+    log.debug('The app has been installed successfully.');
+  } catch (e) {
+    // Want to clarify the device's application installation state in this situation.
+
+    // If the error was by below error case, we could recover the situation
+    // by uninstalling the device's app bundle id explicitly regard less the app exists on the device or not (e.g. offload app).
+    // [XCUITest] Error installing app '/path/to.app': Unexpected data: {"Error":"MismatchedApplicationIdentifierEntitlement","ErrorDescription":"Upgrade's application-identifier entitlement string (TEAM_ID.com.kazucocoa.example) does not match installed application's application-identifier string (ANOTHER_TEAM_ID.com.kazucocoa.example); rejecting upgrade."}
+    if (skipUninstall && e.message.includes('MismatchedApplicationIdentifierEntitlement')) {
+      log.info(`The application identified by '${bundleId}' cannot be installed because it might ` +
+        `be already cached on the device, probably with a different signature. ` +
+        `Will try to remove it and install a new copy. Original error: ${e.message}`);
+      await device.remove(bundleId);
+      await device.install(app, timeout, strategy);
+      log.debug('The app has been installed after one retrial.');
+      return;
+    }
+
+    // Other error cases that could not be recoverable by here.
+    // Exact error will be in the log.
+
+    // We cannot recover 'ApplicationVerificationFailed' situation since this reason is clearly the app's provisioning profile was invalid.
+    // [XCUITest] Error installing app '/path/to.app': Unexpected data: {"Error":"ApplicationVerificationFailed","ErrorDetail":-402620395,"ErrorDescription":"Failed to verify code signature of /path/to.app : 0xe8008015 (A valid provisioning profile for this executable was not found.)"}
+    throw e;
```

Comment:
I would probably put this part into a guard condition instead as it results into a shorter block:

```
if (!skipUninstall || !e.message.includes('MismatchedApplicationIdentifierEntitlement') {
  throw e;
}
....
```